rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+based on a containerized application is no different from any other custom
+component. Below are illustrations to define the components from the previous
+section in full, and their use in an end-to-end pipeline.
+
+*   Component definitions
+
+```python
+
+class MyContainerBasedExampleGen(BaseComponent):
+  """Wraps my_example_gen.py."""
+  SPEC_CLASS = MyContainerBasedExampleGenSpec
+
+  EXECUTOR_SPEC = my_example_gen_exec_sepc
+
+  # Optionally, if custom driver behavior is desired, such as checking
+  # mtime for file updates, one can define a custom Driver class to control
+  # the behavior of my_exmaple_gen.py inside the container.
+  DRIVER_CLASS = ...
+
+
+class MyContainerBasedTrainer(BaseComponent):
+ """Wraps my_trainer.py."""
+  SPEC_CLASS = MyContainerBasedTrainerSpec
+
+  EXECUTOR_SPEC = my_trainer_exec_spec
+
+```
+
+*   `pipeline.py`
+
+```python
+
+def create_pipeline():
+  my_csv_file = dsl_utils.external_input(uri="/path/to/csv_file")
+
+  my_example_gen = MyContainerBasedExampleGen(
+      raw_input=my_csv_file, num_csv_column=20)
+  my_trainer = MyContainerBasedTrainer(
+      my_input=example_gen.outputs.examples, train_steps=200)
+
+  return pipeline.Pipeline(
+    pipeline_name = 'my_container_based_pipeline',
+    pipeline_root = 'gs://path/to/root',
+    components = [my_example_gen, my_trainer],
+    ...
+  )
+
+
+# It may be the case that some TfxRunner implementation (the ComponentLauncher
+# thereof) does not have the ability to run a container-based component,
+# in which case, an Exception is raised at the time when the logical pipeline
+# is compiled for execution by the TfxRunner.
+# See the next section of this doc about ComponentLauncher.
+_ = KubeflowDagRunner().run(_create_pipeline())
+
+```
+
+### ComponentLauncher to launch the container-based application
+
+With the introduction of `ExecutorContainerSpec` which does not specify
+`executor_class`, the default `ComponentLauncher` cannot execute the
+container-based component. Furthermore, different orchestrator (i.e. an instance
+of `TfxRunner`) may have different ways to launch the containerized application
+program.
+
+We propose to extend the base `ComponentLauncher` to define
+orchestrator-specific ways to execute the containerized program, including the
+ways to translate input artifacts to the complete command line, by filling the
+[Jinja template](https://jinja.palletsprojects.com/en/2.10.x/) for
+`ExecutorContainerSpec.command` and `ExecutorContainerSpec.args`, and to
+translate output from the containerized application to keep track of metadata of
+it and write back to Metadata storage.
+
+Below is one possible implementation of `ComponentLauncher` that implements a
+way to launch container-based components with `KubeflowDagRunner`, with ability
+to configure low level k8s configurations. This `KubeflowComponentLauncher`
+would use the k8s Pod API to launch the container [ref]. On top of this,
```

Comment:
Why Pod API? As a Kubernetes user how am I supposed to reason about what's happening? Pods are not resilient. So there needs to be another layer like a K8s Job that restarts failed pods.

Where are the semantics defined? Are  you defining new semantics or using well defined semantics; e.g. an appropriate Kubernetes resource.

Reply:
This will be no different than how `KubeflowDagRunner` launches the default container in k8s ([code](https://github.com/tensorflow/tfx/blob/0.14.0/tfx/orchestration/kubeflow/base_component.py#L118)).  The actual invocation is further abstracted inside KFP sdk (`kfp.dsl.ContainerOp`), which is an implementation of `KubeflowDagRunner` .

The `KubeflowDagRunner` also implements an API to apply KFP-specific modification on the ContainerOp ([code](https://github.com/tensorflow/tfx/blob/master/tfx/orchestration/kubeflow/kubeflow_dag_runner.py#L135)), but is abstracted in the `KubeflowDagRunner` so that logical pipeline construction with the DSL (the `create_pipeline()`) does not see it.